### PR TITLE
DART implementation Knuth Morris Pratt Algorithm

### PR DESCRIPTION
**Pull request by akashgkrishnan**
Implemented a new file with tests for Knuth Morris pratt algorithm
Added test cases both positive and negative

# Welcome to Dart community

### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Dart/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Dart files are placed inside an existing directory.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by ayaankhan98 — COMMENTED** [review]
Thanks for contribution! would you please add relevant documentation to code.
[Documentation Guidelines](https://dart.dev/guides/language/effective-dart/documentation)

**Comment by akashgkrishnan** [comment]
> Thanks for contribution! would you please add relevant documentation to code.
> [Documentation Guidelines](https://dart.dev/guides/language/effective-dart/documentation)

Sure @ayaankhan98 . I'll be adding in the required requested changes

**Comment by Parowicz** [comment]
How it should behave when someone is providing empty `subString`? Right now it crashes.

```dart
  test(('Search empty substring returns <???>: '), () {
    string = 'adafccfefbbbfeeccbcfd';
    subString = '';
    expect(stringCompare(string, subString), isFalse);
  });
```

File: strings/knuth_morris_prat.dart [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+void main() {
+  String string;
+  String subString;
+
+  test(('KMP: '), () {
+    string = 'aefoaefcdaefcdaed';
+    subString = 'aefcdaed';
+    expect(stringCompare(string, subString), equals(true));
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:60]: You can use `isTrue`/`isFalse` instead of `equals(true)`/`equals(false)`
```suggestion
    expect(stringCompare(string, subString), isTrue);
```

File: strings/knuth_morris_prat.dart [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  test(('KMP: '), () {
+    string = 'testwafwafawfawfawfawfawfawfawfa';
+    subString = 'fawfawfawfawfa';
+    expect(stringCompare(string, subString), equals(true));
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:66]:
```suggestion
    expect(stringCompare(string, subString), isTrue);
```

File: strings/knuth_morris_prat.dart [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  test(('KMP: '), () {
+    string = 'aabc';
+    subString = 'abc';
+    expect(stringCompare(string, subString), equals(true));
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:72]:
```suggestion
    expect(stringCompare(string, subString), isTrue);
```

File: strings/knuth_morris_prat.dart [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  test(('KMP: '), () {
+    string = 'adafccfefbbbfeeccbcfd';
+    subString = 'ecb';
+    expect(stringCompare(string, subString), equals(false));
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:78]:
```suggestion
    expect(stringCompare(string, subString), isFalse);
```

File: strings/knuth_morris_prat.dart [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  test(('KMP: '), () {
+    string = 'akash';
+    subString = 'christy';
+    expect(stringCompare(string, subString), equals(false));
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:84]:
```suggestion
    expect(stringCompare(string, subString), isFalse);
```

**Comment by akashgkrishnan** [comment]
@Parowicz @ayaankhan98 hi sir,
please let me know if further changes are required

File: strings/knuth_morris_prat.dart [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+///
+/// Build a pattern which tells us where to continue iteration from if we
+///  get a mismatch between the character
+///
+/// Step through the text one character at a time and compare it to a character in
+/// the pattern updating our location within the pattern if necessary
+bool stringCompare(String string, String subString) {
+  if (subString.isEmpty || string.isEmpty) {
+    return true;
+  }
```
Comment (Parowicz) [inline strings/knuth_morris_prat.dart:14]: I don't know if it's good to return `True` if we are searching something in empty string.
```dart
  test(('Searching in empty string'), () {
    string = '';
    subString = 'asd';
    expect(stringCompare(string, subString), isFalse);
  });
```

Comment (ayaankhan98) [reply strings/knuth_morris_prat.dart:14]: i think it is fine, because this is the way of KMP algorithm. For a empty string there is not pattern in this empty string which is similar to subtring `asd` thus returing false is fine.

Comment (Parowicz) [reply strings/knuth_morris_prat.dart:14]: You mean that test is correct? Because right now it's failing, when `string.isEmpty` `true` is always returned, even when `subString` is not empty.

Comment (ayaankhan98) [reply strings/knuth_morris_prat.dart:14]: Yup i mean that this test is correct, if this test is failing it means the implementaion of KMP algorithm is not correct, it is buggy. returned result should be `false` in this case.

**Comment by akashgkrishnan** [comment]
@Parowicz  @ayaankhan98 I have updated the wrong test case and fixed the edge case. please let me know if this is alright

**Review by Parowicz — APPROVED** [review]
Looks Great! Thank You!

**Review by ayaankhan98 — APPROVED** [review]
LGTM
